### PR TITLE
perf: do not use regex to escape expressions

### PR DESCRIPTION
**Pull request by gurgunday**
See https://github.com/gurgunday/ghtml/pull/16#issuecomment-2156104795

closes https://github.com/gurgunday/ghtml/issues/15

### CODE REVIEW

**Comment by gurgunday** [comment]
Interesting, on macOS, the PR is faster by a big margin

**Comment by gurgunday** [comment]
Kitajs benchmark:

OLD:

```sh
benchmark        time (avg)             (min … max)       p75       p99      p999
--------------------------------------------------- -----------------------------
• Real World Scenario
--------------------------------------------------- -----------------------------
KitaJS/Html     544 µs/iter     (381 µs … 4'877 µs)    422 µs  1'885 µs  4'802 µs
Typed Html    1'950 µs/iter   (1'655 µs … 2'773 µs)  2'200 µs  2'633 µs  2'773 µs
VHtml         2'527 µs/iter   (2'297 µs … 3'608 µs)  2'768 µs  3'517 µs  3'608 µs
React JSX     6'580 µs/iter  (5'995 µs … 10'652 µs)  6'859 µs 10'652 µs 10'652 µs
Preact          966 µs/iter     (688 µs … 3'018 µs)    800 µs  2'451 µs  3'018 µs
React         6'598 µs/iter   (5'997 µs … 7'851 µs)  7'210 µs  7'851 µs  7'851 µs
Common Tags   3'297 µs/iter   (2'915 µs … 5'016 µs)  3'190 µs  4'954 µs  5'016 µs
Ghtml           387 µs/iter     (308 µs … 2'539 µs)    348 µs  2'295 µs  2'503 µs
JSXTE         5'169 µs/iter   (4'961 µs … 9'826 µs)  5'191 µs  5'999 µs  9'826 µs

summary for Real World Scenario
  Ghtml
   1.41x faster than KitaJS/Html
   2.5x faster than Preact
   5.04x faster than Typed Html
   6.52x faster than VHtml
   8.51x faster than Common Tags
   13.35x faster than JSXTE
   16.99x faster than React JSX
   17.04x faster than React

• Component Creation
--------------------------------------------------- -----------------------------
KitaJS/Html     651 µs/iter     (505 µs … 2'261 µs)    529 µs  1'781 µs  2'261 µs
Typed Html    1'017 µs/iter     (823 µs … 2'580 µs)    875 µs  2'250 µs  2'580 µs
VHtml         1'815 µs/iter   (1'536 µs … 3'295 µs)  1'855 µs  3'093 µs  3'295 µs
React JSX     9'184 µs/iter  (8'190 µs … 14'233 µs)  9'549 µs 14'233 µs 14'233 µs
Preact          818 µs/iter     (654 µs … 3'283 µs)    769 µs  2'135 µs  3'283 µs
React         8'997 µs/iter   (8'179 µs … 9'858 µs)  9'424 µs  9'858 µs  9'858 µs
Common Tags   1'845 µs/iter   (1'723 µs … 2'436 µs)  2'021 µs  2'334 µs  2'436 µs
Ghtml           646 µs/iter     (628 µs … 1'098 µs)    633 µs  1'044 µs  1'098 µs
JSXTE         4'733 µs/iter  (4'351 µs … 10'337 µs)  4'749 µs  6'071 µs 10'337 µs

summary for Component Creation
  Ghtml
   1.01x faster than KitaJS/Html
   1.27x faster than Preact
   1.57x faster than Typed Html
   2.81x faster than VHtml
   2.86x faster than Common Tags
   7.32x faster than JSXTE
   13.92x faster than React
   14.21x faster than React JSX

• Attributes Serialization
--------------------------------------------------- -----------------------------
KitaJS/Html  20'114 ns/iter    (18'833 ns … 508 µs) 19'417 ns 24'208 ns    257 µs
Typed Html   79'511 ns/iter    (74'416 ns … 498 µs) 76'041 ns    298 µs    380 µs
VHtml        91'537 ns/iter    (87'000 ns … 528 µs) 88'541 ns    148 µs    381 µs
React JSX    77'500 ns/iter    (70'708 ns … 567 µs) 74'750 ns    192 µs    423 µs
Preact       19'032 ns/iter    (17'458 ns … 515 µs) 18'167 ns 35'208 ns    274 µs
React        77'153 ns/iter    (70'833 ns … 528 µs) 74'125 ns    198 µs    421 µs
Common Tags  52'154 ns/iter    (48'166 ns … 593 µs) 50'541 ns    100 µs    362 µs
Ghtml        46'792 ns/iter    (44'083 ns … 490 µs) 46'125 ns 89'333 ns    329 µs
JSXTE        29'756 ns/iter    (25'792 ns … 517 µs) 27'375 ns 82'833 ns    400 µs

summary for Attributes Serialization
  Preact
   1.06x faster than KitaJS/Html
   1.56x faster than JSXTE
   2.46x faster than Ghtml
   2.74x faster than Common Tags
   4.05x faster than React
   4.07x faster than React JSX
   4.18x faster than Typed Html
   4.81x faster than VHtml
```

NEW:

```sh
benchmark        time (avg)             (min … max)       p75       p99      p999
--------------------------------------------------- -----------------------------
• Real World Scenario
--------------------------------------------------- -----------------------------
KitaJS/Html     507 µs/iter     (389 µs … 1'836 µs)    412 µs  1'221 µs  1'557 µs
Typed Html    1'901 µs/iter   (1'649 µs … 2'670 µs)  2'138 µs  2'600 µs  2'670 µs
VHtml         2'480 µs/iter   (2'301 µs … 3'487 µs)  2'732 µs  3'236 µs  3'487 µs
React JSX     6'459 µs/iter   (6'014 µs … 7'654 µs)  6'881 µs  7'654 µs  7'654 µs
Preact          972 µs/iter     (689 µs … 2'777 µs)    817 µs  2'479 µs  2'777 µs
React         6'597 µs/iter   (6'049 µs … 8'109 µs)  6'885 µs  8'109 µs  8'109 µs
Common Tags   3'078 µs/iter   (2'870 µs … 4'420 µs)  3'051 µs  4'276 µs  4'420 µs
Ghtml           221 µs/iter     (183 µs … 1'454 µs)    203 µs  1'184 µs  1'390 µs
JSXTE         4'404 µs/iter   (3'738 µs … 6'206 µs)  4'458 µs  5'468 µs  6'206 µs

summary for Real World Scenario
  Ghtml
   2.29x faster than KitaJS/Html
   4.39x faster than Preact
   8.59x faster than Typed Html
   11.2x faster than VHtml
   13.91x faster than Common Tags
   19.9x faster than JSXTE
   29.18x faster than React JSX
   29.81x faster than React

• Component Creation
--------------------------------------------------- -----------------------------
KitaJS/Html     563 µs/iter     (505 µs … 1'178 µs)    521 µs    977 µs  1'107 µs
Typed Html      907 µs/iter     (824 µs … 1'535 µs)    853 µs  1'370 µs  1'535 µs
VHtml         1'645 µs/iter   (1'533 µs … 2'875 µs)  1'687 µs  2'500 µs  2'875 µs
React JSX     9'016 µs/iter  (8'177 µs … 10'329 µs)  9'496 µs 10'329 µs 10'329 µs
Preact          770 µs/iter     (656 µs … 1'791 µs)    706 µs  1'550 µs  1'791 µs
React         9'140 µs/iter  (8'233 µs … 12'365 µs)  9'555 µs 12'365 µs 12'365 µs
Common Tags   1'841 µs/iter   (1'724 µs … 2'658 µs)  1'988 µs  2'364 µs  2'658 µs
Ghtml           447 µs/iter       (430 µs … 932 µs)    435 µs    824 µs    917 µs
JSXTE         4'470 µs/iter   (4'274 µs … 5'194 µs)  4'511 µs  4'945 µs  5'194 µs

summary for Component Creation
  Ghtml
   1.26x faster than KitaJS/Html
   1.72x faster than Preact
   2.03x faster than Typed Html
   3.68x faster than VHtml
   4.12x faster than Common Tags
   10x faster than JSXTE
   20.17x faster than React JSX
   20.45x faster than React

• Attributes Serialization
--------------------------------------------------- -----------------------------
KitaJS/Html  20'126 ns/iter    (18'708 ns … 496 µs) 19'375 ns 26'083 ns    316 µs
Typed Html   79'555 ns/iter    (74'250 ns … 429 µs) 75'875 ns    312 µs    399 µs
VHtml        92'170 ns/iter    (87'083 ns … 459 µs) 88'708 ns    179 µs    410 µs
React JSX    78'670 ns/iter    (69'750 ns … 827 µs) 74'750 ns    277 µs    510 µs
Preact       19'414 ns/iter    (17'458 ns … 468 µs) 18'375 ns 46'667 ns    294 µs
React        75'486 ns/iter    (69'875 ns … 430 µs) 73'291 ns    158 µs    358 µs
Common Tags  52'608 ns/iter    (48'458 ns … 508 µs) 50'708 ns    121 µs    344 µs
Ghtml        29'061 ns/iter    (27'250 ns … 398 µs) 28'500 ns 49'250 ns    292 µs
JSXTE        29'119 ns/iter    (25'833 ns … 415 µs) 27'209 ns 54'875 ns    352 µs

summary for Attributes Serialization
  Preact
   1.04x faster than KitaJS/Html
   1.5x faster than Ghtml
   1.5x faster than JSXTE
   2.71x faster than Common Tags
   3.89x faster than React
   4.05x faster than React JSX
   4.1x faster than Typed Html
   4.75x faster than VHtml
```

**Comment by gurgunday** [comment]
@gunjam what do you think?

Kita benchmarks show significant improvement!

Do you have any suggestions to further improve the results?

**Comment by gurgunday** [comment]
Latest, on battery power, improved attribute serialization:

```sh
benchmark        time (avg)             (min … max)       p75       p99      p999
--------------------------------------------------- -----------------------------
• Real World Scenario
--------------------------------------------------- -----------------------------
KitaJS/Html     798 µs/iter     (661 µs … 2'052 µs)    754 µs  1'665 µs  2'052 µs
Typed Html    3'225 µs/iter   (2'653 µs … 4'592 µs)  3'541 µs  4'288 µs  4'592 µs
VHtml         3'925 µs/iter   (3'679 µs … 5'086 µs)  4'161 µs  5'069 µs  5'086 µs
React JSX    10'143 µs/iter  (9'626 µs … 11'669 µs) 10'622 µs 11'669 µs 11'669 µs
Preact        1'516 µs/iter   (1'091 µs … 4'704 µs)  1'479 µs  3'544 µs  4'704 µs
React        10'428 µs/iter  (9'740 µs … 12'066 µs) 11'188 µs 12'066 µs 12'066 µs
Common Tags   4'912 µs/iter   (4'569 µs … 6'434 µs)  4'862 µs  6'433 µs  6'434 µs
Ghtml           378 µs/iter     (293 µs … 2'675 µs)    337 µs  2'459 µs  2'633 µs
JSXTE         7'626 µs/iter   (5'920 µs … 9'351 µs)  8'017 µs  9'351 µs  9'351 µs

summary for Real World Scenario
  Ghtml
   2.11x faster than KitaJS/Html
   4.01x faster than Preact
   8.53x faster than Typed Html
   10.38x faster than VHtml
   12.99x faster than Common Tags
   20.17x faster than JSXTE
   26.82x faster than React JSX
   27.58x faster than React

• Component Creation
--------------------------------------------------- -----------------------------
KitaJS/Html     956 µs/iter     (803 µs … 2'857 µs)    825 µs  2'185 µs  2'857 µs
Typed Html    1'715 µs/iter   (1'308 µs … 3'552 µs)  1'715 µs  3'122 µs  3'552 µs
VHtml         2'855 µs/iter   (2'441 µs … 4'999 µs)  3'005 µs  4'194 µs  4'999 µs
React JSX    13'815 µs/iter (13'008 µs … 14'869 µs) 14'410 µs 14'869 µs 14'869 µs
Preact        1'207 µs/iter   (1'037 µs … 2'576 µs)  1'185 µs  2'283 µs  2'576 µs
React        14'301 µs/iter (13'081 µs … 16'455 µs) 14'849 µs 16'455 µs 16'455 µs
Common Tags   3'075 µs/iter   (2'731 µs … 3'845 µs)  3'209 µs  3'757 µs  3'845 µs
Ghtml           735 µs/iter     (693 µs … 1'464 µs)    719 µs  1'284 µs  1'464 µs
JSXTE         7'326 µs/iter   (6'691 µs … 8'603 µs)  7'793 µs  8'603 µs  8'603 µs

summary for Component Creation
  Ghtml
   1.3x faster than KitaJS/Html
   1.64x faster than Preact
   2.33x faster than Typed Html
   3.88x faster than VHtml
   4.18x faster than Common Tags
   9.97x faster than JSXTE
   18.8x faster than React JSX
   19.46x faster than React

• Attributes Serialization
--------------------------------------------------- -----------------------------
KitaJS/Html  32'302 ns/iter    (29'917 ns … 640 µs) 31'166 ns 43'417 ns    341 µs
Typed Html      130 µs/iter       (123 µs … 582 µs)    126 µs    398 µs    499 µs
VHtml           147 µs/iter       (138 µs … 548 µs)    145 µs    218 µs    531 µs
React JSX       121 µs/iter       (114 µs … 630 µs)    118 µs    210 µs    511 µs
Preact       35'574 ns/iter  (27'958 ns … 2'226 µs) 32'750 ns 71'833 ns    591 µs
React           121 µs/iter       (113 µs … 662 µs)    118 µs    200 µs    522 µs
Common Tags  81'023 ns/iter    (77'000 ns … 528 µs) 79'708 ns 93'542 ns    465 µs
Ghtml        42'918 ns/iter    (41'166 ns … 487 µs) 42'000 ns 47'083 ns    387 µs
JSXTE        47'605 ns/iter    (41'000 ns … 756 µs) 47'208 ns 80'750 ns    526 µs

summary for Attributes Serialization
  KitaJS/Html
   1.1x faster than Preact
   1.33x faster than Ghtml
   1.47x faster than JSXTE
   2.51x faster than Common Tags
   3.73x faster than React JSX
   3.76x faster than React
   4.02x faster than Typed Html
   4.55x faster than VHtml
gurgunday@Gurgun-MacBook-Air html-1 % 
```

**Comment by gunjam** [comment]
ill have a closer look in the morning im exhausted and about to fall asleep (typing this from my steam deck in bed), but looks good!

File: src/html.js [hunk 1]
```diff
@@ -8,11 +8,24 @@ const escapeDictionary = {
...
 
-const escapeFunction = (key) => {
-  return escapeDictionary[key];
+const escapeFunction = (string) => {
+  const stringLength = string.length;
+  let start = 0;
+  let end = 0;
+  let escaped = "";
+
+  do {
```
Comment (gurgunday) [inline src/html.js:20]: no need for a  `for` loop, we can start directly as we only pass a non empty string

File: src/html.js [hunk 2]
```diff
@@ -8,11 +8,24 @@ const escapeDictionary = {
...
+    const escapedCharacter = escapeDictionary[string[end++]];
```
Comment (gurgunday) [inline src/html.js:21]: compiler hint: ++ forces an exp to be a Number and V8 takes note

File: src/html.js [hunk 3]
```diff
@@ -8,11 +8,24 @@ const escapeDictionary = {
...
+    if (escapedCharacter) {
+      escaped += string.slice(start, end - 1) + escapedCharacter;
```
Comment (gurgunday) [inline src/html.js:23]: - 1 is faster than + 1 as `-` operator is not overloaded like `+`

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by gunjam** [comment]
Hey! looks great, nice work!

Also seeing good improvements on my first pass component code with your changes:

main
```
benchmark                         time (avg)             (min … max)       p75       p99      p999
-------------------------------------------------------------------- -----------------------------
govukButton - start now link     306 ns/iter       (268 ns … 624 ns)    313 ns    372 ns    624 ns
govukButton - input              333 ns/iter       (297 ns … 423 ns)    344 ns    390 ns    423 ns
govukButton - button             358 ns/iter       (338 ns … 438 ns)    366 ns    396 ns    438 ns
```
this pr
```
benchmark                         time (avg)             (min … max)       p75       p99      p999
-------------------------------------------------------------------- -----------------------------
govukButton - start now link     194 ns/iter       (171 ns … 364 ns)    198 ns    235 ns    339 ns
govukButton - input              212 ns/iter       (193 ns … 295 ns)    221 ns    252 ns    291 ns
govukButton - button             268 ns/iter       (240 ns … 346 ns)    279 ns    312 ns    345 ns
```

I can't see how you could make it any better, thanks again for your hard work

**Pull request merged**